subprojects/gradle/cd/src/gradleTest/kotlin/com/avito/ci/steps/BuildStepOverrideTest.kt

```diff
@@ -0,0 +1,176 @@
+package com.avito.ci.steps
+
+import com.avito.ci.explicitBuildStepsOverrideProperty
+import com.avito.test.gradle.TestProjectGenerator
+import com.avito.test.gradle.TestResult
+import com.avito.test.gradle.gradlew
+import com.avito.test.gradle.module.AndroidAppModule
+import com.avito.test.gradle.plugin.plugins
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+import java.nio.file.Path
+
+internal class BuildStepOverrideTest {
+
+    private lateinit var projectDir: File
+
+    @BeforeEach
+    fun setup(@TempDir tempPath: Path) {
+        projectDir = tempPath.toFile()
+    }
+
+    @Test
+    fun `register with the same name - overrides previous one - disabled explicit override mode`() {
+        generateProject(
+            """
+            customTask("step") {
+                tasksPredicate = com.avito.ci.TasksPredicate.byName("tasks")
+            }
+            customTask("step") {
+                tasksPredicate = com.avito.ci.TasksPredicate.byName("help")
+            }
+        """.trimIndent()
+        )
+
+        val buildResult = runBuild(
+            explicitOverride = false
+        )
+
+        buildResult.assertThat()
+            .buildSuccessful()
+            .tasksShouldBeTriggered(":app:help", ":app:release")
+
+        buildResult.assertThat()
+            .tasksShouldNotBeTriggered(":app:tasks")
+    }
+
+    @Test
+    fun `register with the same name - fails`() {
+        generateProject(
+            """
+            customTask("step") {
+                tasksPredicate = com.avito.ci.TasksPredicate.byName("tasks")
+            }
+            customTask("step") {
+                tasksPredicate = com.avito.ci.TasksPredicate.byName("help")
+            }
+        """.trimIndent()
+        )
+
+        val buildResult = runBuild(
+            expectFailure = true
+        )
+
+        buildResult.assertThat()
+            .buildFailed()
+            .outputContains("Overriding existing build step 'step'")
+    }
+
+    @Test
+    fun `override by type - overrides previous one`() {
+        generateProject(
+            """
+            customTask("step") {
+                tasksPredicate = com.avito.ci.TasksPredicate.byName("tasks")
+            }
+            overrideStep(com.avito.ci.steps.CustomTaskStep.class) {
```

Comment:
I expect to see `overrideStep(com.avito.ci.steps.CustomTaskStep.class, "step")`

Reply:
Made it a bit more strict in another way. Added a required type to a function with a name.

Now client have two options:
- `overrideStep("name", type)`
- `overrideStep(type)`

`overrideStep("name", type)` is applicable for custom steps when a client knows the name:

```kotlin
customTask("help") {
    tasksPredicate = com.avito.ci.TasksPredicate.byName("help")
}
overrideStep("help", com.avito.ci.steps.CustomTaskStep.class) {
    tasksPredicate = com.avito.ci.TasksPredicate.byName("tasks")
}
```

`overrideStep(type)` is for cases with built-in steps when a client doesn't know names:

```kotlin

// First configuration
getByName("githubIntegrationCheck") {
    uiTests {
        configurations = mutableListOf(Instrumentation.configurations.screenshotNoFlaky)
    }
}

// Override
getByName("githubIntegrationCheck") {
    overrideStep<UiTestCheck> {
        configurations = mutableListOf(Instrumentation.configurations.screenshotNoFlaky)
    }
}
```
In this case, the built-in step (UiTestCheck in this case) can be only one by design. 
To be sure, `overrideStep` check this contract.